### PR TITLE
feat(pie-icons-webc): DSW-984 Convert pie-icons-webc from native web components to Lit

### PR DESCRIPTION
**Pull request by xander-marjoram**
Recommended reviewing using dev mode (press `.`) due to the large number of icon files being deleted!

>**Note**
> There is an existing issue with the webc icons where they will not apply custom classes on top of the base classes. I've raised a separate ticket to cover this (DSW-1034).

## Describe your changes (can list changeset entries if preferable)
- [Fixed] - Invisible icons
- [Fixed] - Prevent tree-shaking of components in storybook
- [Removed] - Built webc icons from source control
- [Added] - Type declaration files for components
- [Added] - Types for pie-icons
- [Added] - TS version of pie-icons-configs/config.js (Will be used after DSW-1025)
- [Added] - Webc icon tests for width, height and base classes
- [Changed] - Update pie-icons-webc build to generate a slightly different template for regular and large icons (using different types, etc.)
- [Changed] - Update pie-icons-webc rollup config to remove commonjs build
- [Changed] - Use `just-kebab-case` and `just-pascal-case` instead of `kebab-case` and `pascal-case` to simplify usage (and they're more recently maintained)

## Author Checklist (complete before requesting a review)
- [x] I have performed a self-review of my code
- [x] If it is a core feature, I have added thorough tests
- [ ] If it is a `PIE Docs` change, I have reviewed the Docs site preview
- [ ] If it is a component change, I have reviewed the Storybook preview
- [ ] If there are visual test updates, I have reviewed them properly before approving

### Reviewer 1
- [ ] If it is a `PIE Docs` change, I have reviewed the PR preview
- [ ] If there are visual test updates, I have reviewed them

### Reviewer 2
- [ ] If it is a `PIE Docs` change, I have reviewed the PR preview
- [ ] If there are visual test updates, I have reviewed them

### CODE REVIEW

File: packages/components/pie-modal/declaration.d.ts [hunk 1]
```diff
@@ -7,3 +7,5 @@ declare module '*.scss?inline' {
     const content: Record<string, string>;
     export default content;
 }
+
+declare module '@justeattakeaway/pie-modal';
```
Comment (jamieomaguire) [inline packages/components/pie-modal/declaration.d.ts:11]: I think that this is called a 'wildcard module declaration' and is sort of like defining the module, and its API's as `any`.

For the scope of your ticket I think this is absolutely fine, however I would love to help look into this further and see how we can improve these declarations (and if that is even necessary)

File: packages/tools/pie-icons/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
+type Icon = {
+    name: string;
+    contents: string;
+    attrs: object;
+    pathPrefix: string;
+}
+
+declare const icons: Record<string, Icon>;
```
Comment (jamieomaguire) [inline packages/tools/pie-icons/index.d.ts:8]: 🫶

File: packages/tools/pie-icons-configs/configs.ts [hunk 3]
```diff
@@ -0,0 +1,129 @@
+// TODO: DSW-1025 - Use this file instead of the JS file
```
Comment (jamieomaguire) [inline packages/tools/pie-icons-configs/configs.ts:1]: Just commenting that i think this is a fantastic improvement! I'd be keen for us to use TS for these kinds of build scripts and packages going forward!

**Review by jamieomaguire — APPROVED** [review]

File: packages/tools/pie-icons/index.d.ts (hunk 2 above)
Comment (xander-marjoram) [reply packages/tools/pie-icons/index.d.ts:8]: A record is a bit like a C# Dictionary, so this is made up of a series of KVPs where the key is the name of the icon and the value is an object representing the icon. The record can hold all of the icons so it doesn't need to be an array 👍🏻

File: packages/components/pie-modal/declaration.d.ts (hunk 1 above)
Comment (xander-marjoram) [reply packages/components/pie-modal/declaration.d.ts:11]: Ah right, yes. It looks like `declaration.d.ts` is for declaring types for an otherwise JS module, and this shouldn't be necessary if we already have `index.d.ts` signposted in `package.json`. I'll try removing it again for now.

File: packages/tools/pie-icons/index.d.ts (hunk 2 above)
Comment (jamieomaguire) [reply packages/tools/pie-icons/index.d.ts:8]: Oh, it was just a heart emoji 😂 thanks for the summary of Records tho!

File: packages/tools/pie-icons-webc/build.js [hunk 4]
```diff
@@ -1,88 +1,105 @@
...
-    }
+const componentSelector = '${kebabCase(name)}';
 
-    set class (value) {
-        this.setAttribute('class', value);
-    }
+export class ${name} extends LitElement implements IconProps {
+    // The following styles make sure that the icon will be sized correctly
+    static styles = css\`
+        :host-context(pie-icon-button) svg,
```
Comment (LTurns) [inline packages/tools/pie-icons-webc/build.js:38]: I saw @ashleynolan mentioned on his PR that :host-context isn't supported on safari or firefox so guessing this needs to be changed? Could be one for a separate PR though

Comment (xander-marjoram) [reply packages/tools/pie-icons-webc/build.js:38]: Yep good spot! I'll see what Ash comes up with then this can be updated

File: packages/tools/pie-icons-configs/declaration.d.ts [hunk 5]
```diff
@@ -0,0 +1 @@
+declare module '@justeattakeaway/pie-icons-configs';
```
[the remark this answers is not in the document]
Comment (xander-marjoram) [reply packages/tools/pie-icons-configs/declaration.d.ts]: I've removed this file in favour of the `index.d.ts`, which should be sufficient. I'll add something into DSW-1025 to check that it's being handled correctly.

File: packages/tools/pie-icons-webc/tests/icon.test.js [hunk 6]
```diff
@@ -1,30 +1,68 @@
 import { html, fixture, expect } from '@open-wc/testing';
 import { sizeToValueMap, largeIconSizeDefault, regularIconSizeDefault } from '@justeattakeaway/pie-icons-configs';
-import { IconAlcoholFilled } from '../icons/IconAlcoholFilled';
-import { IconAlcoholFilledLarge } from '../icons/IconAlcoholFilledLarge';
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+import { IconAlcoholFilled, IconAlcoholFilledLarge } from '../dist/icons';
```
Comment (sebastienrocks) [inline packages/tools/pie-icons-webc/tests/icon.test.js:4]: Since all icon files have been removed, does this mean we need to run an icons build before we can run unit tests?
If so, we should either document this, or make sure the dist build is triggering before running the tests

Comment (xander-marjoram) [reply packages/tools/pie-icons-webc/tests/icon.test.js:4]: Our `turbo.json` handles the dependency here, so running the tests will trigger the build first 👍🏻

**Review by kevinrodrigues — DISMISSED** [review]

File: packages/tools/pie-icons-webc/build.js (hunk 4 above)
Comment (ashleynolan) [reply packages/tools/pie-icons-webc/build.js:38]: I've raised a separate PR for this already, so don't worry about it as part of this PR 👍

File: packages/tools/pie-icons-webc/README.md [hunk 7]
```diff
@@ -27,11 +27,17 @@ yarn add @justeattakeaway/pie-icons-webc
 
 ```js
 // Only import what you need!
-import '@justeattakeaway/pie-icons-webc/icons/IconAppRestaurant';
+import '@justeattakeaway/pie-icons-webc/icons/IconAppRestaurant'; // Import the typescript version of the icon
+// OR
+import '@justeattakeaway/pie-icons-webc/dist/icons/IconAppRestaurant.js'; // Import the ESM version of the icon
```
Comment (ashleynolan) [inline packages/tools/pie-icons-webc/README.md:32]: I'm guessing you wouldn't need to define the `.js` here unless writing TS in your app? In which case, you'd likely just use the TS import above.

If so, I'd probably remove the `.js` here.

Also, if we were to compile the JS files to the icons directory (instead of the `dist/icons` directory), wouldn't that make the import path simpler (as it would be the same potentially I'd think and get handled by the bundler?)

**Review by ashleynolan — COMMENTED** [review]
Few comments from me

File: apps/pie-storybook/stories/pie-button.stories.ts [hunk 8]
```diff
@@ -3,12 +3,12 @@ import type { StoryObj as Story } from '@storybook/web-components';
...
 } from '@justeattakeaway/pie-button';
+import { IconChevronDown, IconPlusCircle } from '@justeattakeaway/pie-icons-webc';
 import { StoryMeta, SlottedComponentProps } from '../types';
 
-import '@justeattakeaway/pie-icons-webc/icons/IconPlusCircle'; // Register icon-plus-circle
-import '@justeattakeaway/pie-icons-webc/icons/IconChevronDown'; // Register icon-chevron-down
+const keptReferences = [IconChevronDown, IconPlusCircle];
 
-type ButtonProps = SlottedComponentProps<ButtonPropsBase>;
+type ButtonProps = SlottedComponentProps<ButtonPropsBase> & { buttonIcons : Array<string> };
```
[the remark this answers is not in the document]
Comment (xander-marjoram) [reply apps/pie-storybook/stories/pie-button.stories.ts:11]: Sure. Without this, the linter complains that the `defaultArgs` we provide don't match the properties of `ButtonProps`. This line extends that type by adding a new property for the `buttonIcons` array that you added.

File: apps/pie-storybook/stories/pie-modal.stories.ts [hunk 9]
```diff
@@ -1,12 +1,9 @@
 import { html, TemplateResult } from 'lit';
 import { ifDefined } from 'lit/directives/if-defined.js';
 import type { StoryObj as Story } from '@storybook/web-components';
-import '@justeattakeaway/pie-button'; // Register pie-button
-import '@justeattakeaway/pie-icon-button'; // Register pie-icon-button
-import '@justeattakeaway/pie-modal'; // Register pie-modal
-import '@justeattakeaway/pie-icons-webc/icons/IconClose'; // Register icon-close
-import '@justeattakeaway/pie-icons-webc/icons/IconChevronLeft'; // Register icon-chevron-left
-import '@justeattakeaway/pie-icons-webc/icons/IconChevronRight'; // Register icon-chevron-right
+import { PieIconButton } from '@justeattakeaway/pie-icon-button';
```
[the remark this answers is not in the document]
Comment (xander-marjoram) [reply apps/pie-storybook/stories/pie-modal.stories.ts:4]: Yes, that's the gist of it.

Out of interest, I checked what Spectrum do and they also seem to have some of their imports like this: https://github.com/adobe/spectrum-web-components/blob/main/packages/dialog/src/Dialog.ts#L28

File: packages/components/pie-modal/src/index.ts [hunk 10]
```diff
@@ -8,10 +8,10 @@ import {
 } from '@justeattakeaway/pie-webc-core';
 import type { DependentMap } from '@justeattakeaway/pie-webc-core';
 import { disableBodyScroll, enableBodyScroll } from 'body-scroll-lock';
-import '@justeattakeaway/pie-icons-webc/icons/IconClose';
-import '@justeattakeaway/pie-icons-webc/icons/IconChevronLeft';
-import '@justeattakeaway/pie-icons-webc/icons/IconChevronRight';
 import { Variant } from '@justeattakeaway/pie-button/src/defs.ts';
+import '@justeattakeaway/pie-icons-webc/dist/icons/IconClose.js';
```
[the remark this answers is not in the document]
Comment (xander-marjoram) [reply packages/components/pie-modal/src/index.ts:12]: I did try that initially but the named import goes down a bit of a rabbit hole of eslint and tslint ignore statements!

File: packages/tools/pie-icons/index.d.ts [hunk 11]
```diff
@@ -0,0 +1,8 @@
+type Icon = {
```
[the remark this answers is not in the document]
Comment (xander-marjoram) [reply packages/tools/pie-icons/index.d.ts:1]: Good spot, I'll add that now.

File: packages/tools/pie-icons-webc/README.md (hunk 7 above)
Comment (xander-marjoram) [reply packages/tools/pie-icons-webc/README.md:32]: I was actually leaning the other way and considered removing the TS instructions!
I tried to import the TS versions in pie-modal but that was breaking on CI, saying it [couldn't resolve the import](https://github.com/justeattakeaway/pie/actions/runs/5712859793/job/15477451479#step:5:135). The JS works just fine.

File: packages/tools/pie-icons-webc/icons/IconAlcohol.js [hunk 12]
```diff
@@ -1,61 +0,0 @@
-import { getSvgProps } from '@justeattakeaway/pie-icons-configs';
```
[the remark this answers is not in the document]
Comment (xander-marjoram) [reply packages/tools/pie-icons-webc/icons/IconAlcohol.js:1]: The current tests do check that there is an SVG element inside the icon component, and that it has the correct width and height. I'm not sure a snapshot test would give us much more?

File: packages/components/pie-modal/src/index.ts [hunk 13]
```diff
@@ -350,7 +351,7 @@ export class PieModal extends RtlMixin(LitElement) implements ModalProps {
             ?isFullWidthBelowMid=${isFullWidthBelowMid}
             ?isLoading=${isLoading}
             aria-busy="${isLoading ? 'true' : 'false'}"
-            aria-label="${isLoading ? aria?.loading : nothing}"
+            aria-label="${isLoading ? (aria?.loading || nothing) : nothing}"
```
[the remark this answers is not in the document]
Comment (xander-marjoram) [reply packages/components/pie-modal/src/index.ts:354]: I think this can actually be simplified further to `(isLoading && aria?.loading) || nothing`. Personally I think they can all be a bit tricky to read but hopefully with the shorter one there is less that needs to be broken down 😄

File: apps/pie-storybook/stories/pie-button.stories.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (raoufswe) [reply apps/pie-storybook/stories/pie-button.stories.ts:11]: 📓 no action is required for the following but more of a knowledge sharing.

the following suggestion will probably create the same thing but considering `SlottedComponentProps` is generic that should wrap _all_ types, I'd have extended the generic value instead of the whole type.

```suggestion
type ButtonProps = SlottedComponentProps<ButtonPropsBase & { buttonIcons : Array<string> }>;
```

also at this point, I'm not sure about the value of having `SlottedComponentProps` when the following would be a lot easier to customize.

```suggestion
type ButtonProps = ButtonPropsBase & { slot: string; buttonIcons : Array<string> };
```

we shouldn't keep adding generics for each property we want to add as this is usually done through extending interfaces from a base interface or with `&`. generics are useful when we want to handle multiple dynamic types as they provide flexibility of types at compile time.  for example:

```
const fetchData<T> = (data) => {
const res: {
isLoading: boolean;
data: T } =  {
  isLoading: false,
  data: data
}
return res
}

fetchData<string[]>(['Hellow'])
```

File: packages/components/pie-modal/src/index.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (raoufswe) [reply packages/components/pie-modal/src/index.ts]: yes go ahead 🔥

**Review by raoufswe — APPROVED** [review]

File: apps/pie-storybook/stories/pie-modal.stories.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (ashleynolan) [reply apps/pie-storybook/stories/pie-modal.stories.ts:4]: Maybe catch up on this one tomorrow, but think we've got wires crossed. Looks like that Spectrum reference is the same as it was before your change? (which is what would go with if it means you don't need the `keptReferences`)

File: packages/tools/pie-icons-webc/README.md (hunk 7 above)
Comment (ashleynolan) [reply packages/tools/pie-icons-webc/README.md:32]: I was mainly just wondering if we outputted everything to the `icons` folder, if we could just get rid of the `dist` folder entirely. It would make the import statements a bit cleaner then (if that works as expected).

What's the use-case for importing the TS icons out of interest?

File: packages/tools/pie-icons-webc/icons/IconAlcohol.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (ashleynolan) [reply packages/tools/pie-icons-webc/icons/IconAlcohol.js:1]: I think it'd be useful to just have a single snapshot for one of the icons – happy for this to be done as a follow-up ticket though.

File: packages/components/pie-modal/src/index.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (ashleynolan) [reply packages/components/pie-modal/src/index.ts:12]: Ok – I think we'll likely look at some of this when we explore bundling a bit deeper soon, so happy to leave it as it is for now.

File: apps/pie-storybook/stories/pie-modal.stories.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (ashleynolan) [reply apps/pie-storybook/stories/pie-modal.stories.ts:4]: Ok – I think we'll likely look at some of this when we explore bundling a bit deeper soon, so happy to leave it as it is for now.

File: apps/pie-storybook/stories/pie-button.stories.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (xander-marjoram) [reply apps/pie-storybook/stories/pie-button.stories.ts:11]: I think `SlottedComponentProps` is still valuable because there will be a lot of components that take a slot, however only this component (and probably the form input?) will take a list of icons so it didn't make much sense to me to pull that out into another generic wrapper.

In terms of where to put the `>`, I'm not sure it makes too much difference but personally I prefer the way the first one reads

File: packages/tools/pie-icons-webc/icons/IconAlcohol.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (xander-marjoram) [reply packages/tools/pie-icons-webc/icons/IconAlcohol.js:1]: I've raised a ticket for this, DSW-1043
